### PR TITLE
compute: produce `Export` events for logging dataflows

### PR DESCRIPTION
**Pull request by teskje**
This PR arranges for `Export` events to be logged for logging dataflows. Previously, these events were only logged for dataflows created through `handle_create_dataflow`, which I think was an oversight. At least various places in the logic of the compute logging dataflow assume that an `Export` event is emitted for each dataflow before any other events are emitted.

### Motivation

  * This PR fixes a previously unreported bug.

`ComputeEvent::Export` is never omitted for logging dataflows.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - N/A

### CODE REVIEW

File: src/compute/src/compute_state.rs [hunk 1]
```diff
@@ -501,15 +501,12 @@ impl<'a, A: Allocate + 'static> ActiveComputeState<'a, A> {
...
+        for (log, (trace, dataflow_index)) in traces {
+            // Install trace as maintained index.
             let id = config.index_logs[&log];
             self.compute_state.traces.set(id, trace);
-        }
 
-        // Initialize compute and logging state for each logging index.
-        let index_ids = config.index_logs.values().copied();
-        for id in index_ids {
+            // Initialize compute and logging state for the logging index.
```
Comment (antiguru) [inline src/compute/src/compute_state.rs:509]: This removes an implicit check that we have a log for all `config.index_logs`. If it's something we want to keep, we should make it explicit!

File: src/compute/src/logging/initialize.rs [hunk 2]
```diff
@@ -88,31 +91,40 @@ struct LoggingContext<'a, A: Allocate> {
 }
 
 impl<A: Allocate + 'static> LoggingContext<'_, A> {
-    fn construct_dataflows(&mut self) -> BTreeMap<LogVariant, TraceBundle> {
-        let mut traces = BTreeMap::new();
-        traces.extend(super::timely::construct(
+    fn construct_dataflows(&mut self) -> BTreeMap<LogVariant, (TraceBundle, usize)> {
+        let mut dataflows = BTreeMap::new();
+
+        let index = self.worker.next_dataflow_index();
```
Comment (antiguru) [inline src/compute/src/logging/initialize.rs:97]: This assumes that each `construct` call only creates a single dataflow. Would it be nicer if `construct` returned `(LogVariant, (usize, RowRowAgent<...>, Rc<...>))` to encode the dataflow id per log variant?

**Review by antiguru — COMMENTED** [review]
Looks fine, leaving some questions about further improvements we could do!

File: src/compute/src/compute_state.rs (hunk 1 above)
Comment (teskje) [reply src/compute/src/compute_state.rs:509]: I was wondering about this as well. The previous code would make sure that a `CollectionState` and logging was added for all logs in the config, but it didn't actually check to ensure that we also have a trace for each of them. That doesn't seem very useful: I'm pretty sure if anyone would actually want to query a collection that has no known trace, we would just panic anyway.

Probably doesn't hurt to add a check here that ensures all requested variants have a trace though.

File: src/compute/src/logging/initialize.rs (hunk 2 above)
Comment (teskje) [reply src/compute/src/logging/initialize.rs:97]: Maybe! Probably a good opportunity to introduce a struct to replace the tuple too.

File: src/compute/src/compute_state.rs (hunk 1 above)
Comment (teskje) [reply src/compute/src/compute_state.rs:512]: Added a sanity check.

**Review by antiguru — APPROVED** [review]
Thanks, looks great!

**Comment by teskje** [comment]
TFTR!

**Pull request merged**